### DISCUSSION TITLE
Gracefully stopping slimming process for daemon

### DISCUSSION THREAD

**Discussion by bt-nia** [slimtoolkit/slim#205] [Q&A] [answered]
hey everyone!
I have a docker container that runs as a daemon, which means that it will not stop when running --exec. Before the analysis stops, I'm interacting with the container a few times and then would like the analysis to stop. 
How can I let docker-slim know that the container analysis is done? I've seen the red text telling me to press enter once done before, but for some reason, it does not appear with this container.

**Comment by kcq** [comment, marked as answer]
What kind of container image do you have? Is it a web or API server? Does it have any exposed ports? How did you invoke `docker-slim` and what command flags did you use?

It's possible that what you are seeing is the default `docker-slim` behavior for the container images that expose ports. If that's the case then `docker-slim` will run its probes against the temporary container it creates based on your container image.

The `--continue-after` parameter to `docker-slim` defines how you want to decide when the analysis is done. The parameter defaults to `probe`, which means that `docker-slim` will wait until all (http) probes are done executing and then it moves on to the next phase. 

In addition to configuring custom http probe commands with `--http-probe-cmd` and `--http-probe-cmd-file` you can also invoke external apps during the probing phase. This can be done using the `--http-probe-exec` and `--http-probe-exec-file` parameters, which allows you to use those external apps to decide when the analysis is done.

There are other `--continue-after` options you can use to decide when the analysis phase is done. `enter` tells `docker-slim` to wait for external console input. That way you can manually signal the end of the analysis phase. This is that red text you mentioned.

The `signal` `--continue-after` option lets you use unix signals to let `docker-slim` know when the analysis phase should be done.

The `timeout` `--continue-after` option gives you a time-based approach to decide when the analysis phase is done. You can specify a number to `--continue-after` to choose how many seconds you want to wait before the analysis phase is done.

The exec `--continue-after` option lets you indicate that you want the analysis phase to be done once the shell commands passed through the `--exec` or `--exec-file` params are done executing in the temporary container.

Note that you can combine the `exec` and `probe` options (e.g., `--continue-after probe&exec`). With these stacked options the analysis phase will be done once the exec commands are done executing and the probing phase is done.

**Reply by bt-nia** [reply]
Thanks a lot @kcq for the detailed response that covers all my questions!

**Reply by bt-nia** [reply]
> What kind of container image do you have? Is it a web or API server? Does it have any exposed ports? How did you invoke `docker-slim` and what command flags did you use?
> 
> It's possible that what you are seeing is the default `docker-slim` behavior for the container images that expose ports. If that's the case then `docker-slim` will run its probes against the temporary container it creates based on your container image.
> 
> The `--continue-after` parameter to `docker-slim` defines how you want to decide when the analysis is done. The parameter defaults to `probe`, which means that `docker-slim` will wait until all (http) probes are done executing and then it moves on to the next phase.
> 
> In addition to configuring custom http probe commands with `--http-probe-cmd` and `--http-probe-cmd-file` you can also invoke external apps during the probing phase. This can be done using the `--http-probe-exec` and `--http-probe-exec-file` parameters, which allows you to use those external apps to decide when the analysis is done.
> 
> There are other `--continue-after` options you can use to decide when the analysis phase is done. `enter` tells `docker-slim` to wait for external console input. That way you can manually signal the end of the analysis phase. This is that red text you mentioned.
> 
> The `signal` `--continue-after` option lets you use unix signals to let `docker-slim` know when the analysis phase should be done.
> 
> The `timeout` `--continue-after` option gives you a time-based approach to decide when the analysis phase is done. You can specify a number to `--continue-after` to choose how many seconds you want to wait before the analysis phase is done.
> 
> The exec `--continue-after` option lets you indicate that you want the analysis phase to be done once the shell commands passed through the `--exec` or `--exec-file` params are done executing in the temporary container.
> 
> Note that you can combine the `exec` and `probe` options (e.g., `--continue-after probe&exec`). With these stacked options the analysis phase will be done once the exec commands are done executing and the probing phase is done.

I have a follow-up question: Is it possible to set a timeout for the `exec`? From how I understand it, the two options are always distinct from another.

**Reply by kcq** [reply]
The docker API doesn't provide a way to interrupt an executing `exec` session, so timing out the shell commands from the `exec` flag would have to be done using some kind of wrapper script. Need to investigate the options here.
